### PR TITLE
CRX: make disabling extension during run, optional

### PR DESCRIPTION
**Pull request by ebidel**
R: @brendankenny, all

This makes the enabling/disabling of extension before a LH run a user setting:

screen shot 2017-02-02 at 11 24 53 am

My worry was that some users would find the current default behavior (disable all extension before the run) unexpected and frustrating. Any extensions that don't save state when they're killed off will lose data. I suspect there are many that don't follow best practices. Instead, we're making this a setting for people to opt into.

@brendankenny I tried to move some of the bg methods into popup.html like we chatted about but it was getting trickier than it was worth. We allow the user to dismiss the popup while LH is running. It was getting gnarly to make a transient popup communicate state back to the bg page for persistence, then repopulate the UI when the user reopens the popup. I was also finding myself piping more `background` page variables through popup.js methods...which defeated the original purpose of the code refactor (separation of concerns between bg page and popup.js).

### CODE REVIEW

**Comment by ebidel** [comment]
Tests are shoppy shop. Shop is the worst.

File: lighthouse-extension/app/src/popup.js [hunk 1]
```diff
@@ -207,9 +209,11 @@ function initPopup() {
     const okButton = document.getElementById('ok');
     okButton.addEventListener('click', () => {
       // Save selected aggregation categories on options page close.
```
Comment (brendankenny) [inline lighthouse-extension/app/src/popup.js]: "Save settings on options page close"?

File: lighthouse-extension/app/src/lighthouse-background.js [hunk 2]
```diff
@@ -26,8 +26,10 @@ const log = require('../../../lighthouse-core/lib/log');
 const ReportGenerator = require('../../../lighthouse-core/report/report-generator');
 
 const STORAGE_KEY = 'lighthouse_audits';
+const SETTINGS_KEY = 'lighthouse_settings';
 
 let installedExtensions = [];
+let DISABLE_EXTENSIONS_DURING_RUN = false;
```
Comment (brendankenny) [inline lighthouse-extension/app/src/lighthouse-background.js]: not a constant so shouldn't be all caps

File: lighthouse-extension/app/src/lighthouse-background.js [hunk 3]
```diff
@@ -251,41 +257,57 @@ window.getDefaultAggregations = function() {
 
 /**
  * Save currently selected set of aggregation categories to local storage.
- * @param {!Array<{name: string, audits: !Array<string>}>} selectedAggregations
+ * @param {!{selectedAggregations: Array<{name: string, audits: !Array<string>}>,
+ *           disableExtensions: boolean}} settings
+ * @param {!boolean} disableExtensions True if extensions should be disabled for the run.
```
Comment (brendankenny) [inline lighthouse-extension/app/src/lighthouse-background.js]: not separate params, just one object

**Comment by brendankenny** [comment]
> I tried to move some of the bg methods into popup.html like we chatted about but it was getting trickier than it was worth.

maybe (in a future PR) we can move it into a separate background script, then, so other clients won't get the extension-specific stuff

File: lighthouse-extension/app/src/lighthouse-background.js (hunk 3 above)
Comment (ebidel) [inline lighthouse-extension/app/src/lighthouse-background.js]: leftover from tinkering, done.

File: lighthouse-extension/app/src/lighthouse-background.js (hunk 2 above)
Comment (ebidel) [inline lighthouse-extension/app/src/lighthouse-background.js]: done

**Comment by ebidel** [comment]
> maybe (in a future PR) we can move it into a separate background script, then, so other clients won't get the extension-specific stuff

Or we flip it and have something special for devtools. It's funky the devtools using the built background file from the extension. Those should really be separate targets.

**Comment by ebidel** [comment]
PTAL

**Comment by ebidel** [comment]
Tests are green

File: lighthouse-extension/app/src/lighthouse-background.js [hunk 4]
```diff
@@ -251,41 +257,56 @@ window.getDefaultAggregations = function() {
 
 /**
  * Save currently selected set of aggregation categories to local storage.
- * @param {!Array<{name: string, audits: !Array<string>}>} selectedAggregations
+ * @param {!{selectedAggregations: Array<{name: string, audits: !Array<string>}>,
```
Comment (brendankenny) [inline lighthouse-extension/app/src/lighthouse-background.js]: somewhere along the line this became wrong. I believe its `@param {{selectedAggregations: !Array<string>, disableExtensions: boolean}} settings`

File: lighthouse-extension/app/src/lighthouse-background.js [hunk 5]
```diff
@@ -251,41 +257,56 @@ window.getDefaultAggregations = function() {
...
+  storage[SETTINGS_KEY].disableExtensions = disableExtensionsDuringRun;
+
+  // Save object to chrome local storage.
   chrome.storage.local.set(storage);
 };
 
 /**
  * Load selected aggregation categories from local storage.
- * @return {!Promise<!Object<boolean>>}
+ * @return {!Promise<{selectedAggregations: Array<{name: string, audits: !Array<string>}>,
```
Comment (brendankenny) [inline lighthouse-extension/app/src/lighthouse-background.js]: this one was right, so `@return {!Promise<{selectedAggregations: !Object<boolean>, disableExtensions: boolean}>}`

File: lighthouse-extension/app/src/lighthouse-background.js (hunk 4 above)
Comment (ebidel) [inline lighthouse-extension/app/src/lighthouse-background.js]: done

File: lighthouse-extension/app/src/lighthouse-background.js (hunk 5 above)
Comment (ebidel) [inline lighthouse-extension/app/src/lighthouse-background.js]: done

**Comment by ebidel** [comment]
Woot!

**Pull request merged**